### DISCUSSION TITLE
Allow translations in user generated content

### DISCUSSION THREAD

**Discussion by rijkvanzanten** [directus/directus#3075] [Feature Requests]
Most of the app can be translated, however, there's one glaring omission: user generated values like placeholders, checkbox labels, etc.

It would be great to have a way to inject custom translation values that you then can use in things like interface options, field comments, etc.

I'm thinking a JSON field in directus_settings that holds:

```json
{
  "en-US": {
    "my_custom_placeholder": "Placeholder"
  },
  "<language>": {
    "<key>": "<translation"
  }
}
```

which you can then use in placeholders etc by prefixing the key with a 'translate this value' prefix like `$t:` : 

```json
{
  "options": {
    "placeholder": "$t:my_custom_placeholder"
  }
}
```

**Comment by t7tran** [comment]
I attempted this [Component based localization](https://kazupon.github.io/vue-i18n/guide/component.html) in the custom modules. They work but it seems to cause memory leak. After switching between languages, the Chrome tab freezes, CPU maxes out and Memory is raising to GBs. Any ideas guys?

**Reply by rijkvanzanten** [reply]
No idea about that particular memory issue, no.

That being said, I think we should add a `translations` property to the index of custom extensions so you can provide an object of translations that merged into the global translations pool 🤔

**Reply by t7tran** [reply]
That is a long-awaited feature. I come across a number of questions on Discord.

**Reply by rijkvanzanten** [reply]
Wanna try opening a PR for that? Sounds useful to have!

**Reply by t7tran** [reply]
What would you suggest for a starter? I don't think a JSON field in `directus_settings` is a good place. The similar mechanism/structure in the app/lang folder should be used. I wished it were like:

```
/directus/extension/translations
  /en-US.yaml
  /vi-VN.yaml
```

Then it could be checked for availability before loaded and merged into the main stream as `setLanguage(lang)` invocation.

```
	const translations = await import(`@/lang/translations/${lang}.yaml`).catch((err) => console.warn(err));
	i18n.mergeLocaleMessage(lang, translations);
	// pseudo code
	// i18n.mergeLocaleMessage(lang, await import `/directus/extensions/translations/${lang}.yaml`)...
	loadedLanguages.push(lang);
```

Apparently, this runs in the browser and I have no idea of how dynamic import custom translation YAML can be done yet.

**Reply by rijkvanzanten** [reply]
Ohh I was thinking in the index.js files for custom modules 🤔 Haven't thought through a bigger "global" option for end users yet =

**Reply by t7tran** [reply]
Well, I wouldn't expect less if we start supporting this. We build custom displays, interfaces, layouts, modules together in a single repo. Regardless, all will be loaded in the app eventually. The translations should be reused across all extensions in similar manner as the app.

Bonus: this translations structure allows other teams to utilise the same CrowdIn integration.

**Reply by rijkvanzanten** [reply]
Ahhh that's a very good point, didn't think about that 👍🏻

**Reply by rijkvanzanten** [reply]
This ties in to the same "bundled extensions" discussions as elsewhere (@nickrum )

**Reply by nickrum** [reply]
With the current plan that would probably require us to add a new `translation` extension type.

**Comment by benhaynes** [comment]
Will this also work for adding translations into Extensions? @rijkvanzanten @nickrum

**Reply by rijkvanzanten** [reply]
We would provide a utility that traverses over an object to replace every value that has the `$t:` prefix to use the registered translations. Next to that, we'll add a way for extensions to register new translations themselves 👍🏻

**Comment by benhaynes** [comment]
If you have a LOT of centralized translations it might be nice if you can use `/` or `.` (or any other idea) in the key to organize them into categories in the translation KVP manager. That would make a long list far more organized and digestible.

**Reply by rijkvanzanten** [reply]
That's how it works currently ✨ 

https://github.com/directus/directus/blob/50fdea17099197a493b57a63056c0e7b14bb33f2/app/src/lang/translations/en-US.yaml#L515-L520

w/ `$t:operators.eq`

**Reply by benhaynes** [reply]
Yeah, just thinking we'd want to mirror this in the UI for configuring the KVPs in the App. 👍

**Comment by Pachat** [comment]
To have `Note` and `Placeholder` translations, as we have for `Field Name`,  would be great

**Reply by bcartier** [reply]
This would be a huge improvement.

**Comment by rijkvanzanten** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
